src/buffer/buffer.rs

```diff
@@ -91,19 +91,90 @@ impl Buffer {
+    /// panics instead. This is kept for backwards compatibility. See `get_opt` for a safe
+    /// alternative.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the index is out of bounds.
     #[track_caller]
     pub fn get(&self, x: u16, y: u16) -> &Cell {
         let i = self.index_of(x, y);
         &self.content[i]
     }
 
     /// Returns a mutable reference to Cell at the given coordinates
+    ///
+    /// Callers should generally use the `IndexMut` trait (`&mut buf[(x, y)]`) or the `get_mut_opt`
+    /// method instead of this method.
+    ///
+    /// Note that conventionally methods named `get_mut` usually return `Option<&mut T>`, but this
+    /// method panics instead. This is kept for backwards compatibility. See `get_mut_opt` for a
+    /// safe alternative.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the index is out of bounds.
     #[track_caller]
     pub fn get_mut(&mut self, x: u16, y: u16) -> &mut Cell {
         let i = self.index_of(x, y);
         &mut self.content[i]
     }
 
+    /// Returns a reference to Cell at the given coordinates.
+    ///
+    /// Returns `None` if the index is out of bounds.
+    ///
+    /// Note that unlike `get`, this method accepts a `Position` instead of `x` and `y` coordinates.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use ratatui::{prelude::*, buffer::Cell, layout::Position};
+    /// let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 10));
+    ///
+    /// assert_eq!(buffer.get_opt((0, 0)), Some(&Cell::default()));
+    /// assert_eq!(buffer.get_opt((10, 10)), None);
+    /// assert_eq!(buffer.get_opt(Position::new(0, 0)), Some(&Cell::default()));
+    /// assert_eq!(buffer.get_opt(Position::new(10, 10)), None);
+    /// ```
+    pub fn get_opt<P: Into<Position>>(&self, pos: P) -> Option<&Cell> {
+        let pos = pos.into();
+        let index = self.index_of_opt(pos.x, pos.y)?;
+        self.content.get(index)
+    }
+
+    /// Returns a mutable reference to Cell at the given coordinates
+    ///
+    /// Returns `None` if the index is out of bounds.
+    ///
+    /// Note that unlike `get`, this method accepts a `Position` instead of `x` and `y` coordinates.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use ratatui::{prelude::*, buffer::Cell, layout::Position};
+    /// let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 10));
+    ///
+    /// assert_eq!(buffer.get_mut_opt((0, 0)), Some(&mut Cell::default()));
+    /// assert_eq!(buffer.get_mut_opt((10, 10)), None);
+    /// assert_eq!(
+    ///     buffer.get_mut_opt(Position::new(0, 0)),
+    ///     Some(&mut Cell::default())
+    /// );
+    /// assert_eq!(buffer.get_mut_opt(Position::new(10, 10)), None);
+    /// ```
+    pub fn get_mut_opt<P: Into<Position>>(&mut self, pos: P) -> Option<&mut Cell> {
+        let pos = pos.into();
+        let index = self.index_of_opt(pos.x, pos.y)?;
+        self.content.get_mut(index)
+    }
```

Comment:
Should this function name be `get_mut_opt` or `get_opt_mut`? 

It returns `Option<&mut Cell>` and not `&mut Option<Cell>`, so reading left to right would return in `get_opt_mut`. 

https://github.com/search?q=get_opt_mut+language%3Arust&type=code has 143 results
https://github.com/search?q=get_mut_opt+language%3Arust&type=code has 19 results

Reply:
I think probably reading inside out might make more sense - I'd anticipate something that is suffixed `_mut` to return `&mut something` generally. I'm not 100% sure though. I wouldn't necessarily worry about quantity of results here and instead be more concerned more about being consistent with a reasonably obvious rule.

Reply:
I don't know about the reading inside out? How do you read `Arc<Mutex<T>`? The only part I read inside out in Rust is "mutable reference"; I think everything else I tend to read left to right.

btw, if we introduce `RenderContext`, would we have `RenderContext::get` and `RenderContext::get_mut` return `Option<T>`?

Reply:
Rust has settled on using suffixes to modify the type of thing returned, not prefixes
E.g. https://rust-lang.github.io/rfcs/0199-ownership-variants.html

`_mut` is a modifier to the thing that came before.
- If `get` returns a `Cell`, then `get_mut` returns `&mut Cell`
- If `get_opt` returns an `Option<Cell>` then get_opt_mut returns `&mut Option<Cell>`

More generally when you apply a suffix to an item, you do so in the order from right to left, so IMHO:
- `get_opt_mut` should return &mut<Option<T>>
- `get_mut_opt` should return Option<&mut T>

Regardless, `cell` and `cell_mut` are a couple of better names (or `get_cell`/`get_cell_mut`) that would avoid this altogether.